src/shaders/pbr.frag

```diff
@@ -0,0 +1,274 @@
+// NOTE: In this shader, the light intensity is considered in the lightColor!!
+uniform vec3 LightColors[LIGHT_COUNT];
+uniform float LightRanges[LIGHT_COUNT];
+
+// if .w == 0, it means it is a dirctional light, .xyz is the direction;
+// if .w == 1, it means it is a point light, .xyz is the light position;
+// it is NOT put in the Light Structure, simply because we may modify the code
+// so it is computed in the vertex shader.
+uniform vec4 LightDirections[LIGHT_COUNT];
+
+// -------------- material, textures ------------------
+struct MaterialData {
+  vec4 baseColor;   // diffuse color, if baseTexture exists,
+                    // use the baseTexture
+  float roughness;  // roughness of a surface, if roughness texture exists,
+                    // use the roughnessTexture
+  float metallic;   // metalness of a surface, if metallic texture exists,
+                    // use the metallicTexture
+};
+uniform MaterialData Material;
+
+#if defined(BASECOLOR_TEXTURE)
+uniform sampler2D baseTexture;
+#endif
+#if defined(ROUGHNESS_TEXTURE)
+uniform sampler2D roughnessTexture;
+#endif
+#if defined(METALLIC_TEXTURE)
+uniform sampler2D metallicTexture;
+#endif
+#if defined(NORMAL_TEXTURE)
+uniform sampler2D normalTexture;
+#endif
+
+#if defined(NONE_ROUGHNESS_METALLIC_TEXTURE) || defined(OCCLUSION_ROUGHNESS_METALLIC_TEXTURE)
+uniform sampler2D combinedTexture;
+#endif
+
+// -------------- uniforms ----------------
+#if defined(OBJECT_ID)
+uniform highp uint objectId;
+#endif
+
+#if defined(NORMAL_TEXTURE) && defined(NORMAL_TEXTURE_SCALE)
+uniform mediump float normalTextureScale
+#ifndef GL_ES
+    = 1.0
+#endif
+    ;
+#endif
+
+// -------------- shader ------------------
+#if defined(NORMAL_TEXTURE)
+vec3 getNormalFromNormalMap() {
+  vec3 tangentNormal =
+#if defined(NORMAL_TEXTURE_SCALE)
+      normalize((texture(normalTexture, texCoord).xyz * 2.0 - 1.0) *
+                vec3(normalTextureScale, normalTextureScale, 1.0));
+#else
+      texture(normalTexture, texCoord).xyz * 2.0 - 1.0;
+#endif
+
+#if defined(PRECOMPUTED_TANGENT)
+  mat3 TBN = mat3(tangent, biTangent, normal);
+#else
+  // Normal Mapping Without Precomputed Tangents
+  // See here: http://www.thetenthplanet.de/archives/1180
+  vec3 q1 = dFdx(position);
+  vec3 q2 = dFdy(position);
+  vec2 st1 = dFdx(texCoord);
+  vec2 st2 = dFdy(texCoord);
+
+  vec3 N = normalize(normal);
+  vec3 T = normalize(q1 * st2.t - q2 * st1.t);
+  vec3 B = -normalize(cross(N, T));
+  mat3 TBN = mat3(T, B, N);
+#endif
+
+  return normalize(TBN * tangentNormal);
+}
+
+const float PI = 3.14159265359;
+const float Epsilon = 0.0001;
+
+// Specular D, normal distribution function (NDF),
+// also known as ggxDistribution
+// normal: normal direction
+// light: light source direction
+// viwer: camera direction, aka light outgoing direction
+// half: half vector of light and view
+float normalDistribution(vec3 normal, vec3 half, float roughness) {
+  float a = roughness * roughness;
+  float a2 = a * a;
```

Comment:
Shouldn't all you squaring be using the 'exp2' function? http://csciwww.etsu.edu/barrettm/4157/OpenGL%20Functions.htm